### ISSUE TITLE
Processing GraphQL input variables.

### ISSUE THREAD

**Issue by vpetrovykh** [edgedb/edgedb#1241]
GraphQL has input object types defined in the schema. This means that one could look up what they are and use an variable of one of those types, such as `InsertUser` or even `FilterUser` or `OrderUser`. The first one of these examples mostly contains data, but if `User` has links it may also contain command for filtering objects to be linked as opposed to raw data. The other two are just commands abstracted away into input objects.

We use these commands from input objects to compile the query, so these "variables" aren't really query parameters, they may just *be* the query itself. E.g. `OrderUser` variable could be passed as `{"order": {"name": {"dir": "ASC"}}}`. The final query will have an `ORDER BY .name ASC` clause, but won't have any parameters.

However, the more likely application of passing complex data to a GraphQL query is some kind of bulk insert scenario, where the data is an array of `InsertUser` objects for example. In theory if we restrict the elements to be homogeneous in structure (such that they all fit the same query shape) it would be possible to generate a `FOR ... INSERT` query that uses JSON functions to decompose the original JSON input blob into a bulk insert. 

The bottom line is that the current implementation is limited to dealing with only scalar input variables and doesn't have a way of dealing with the above scenarios. In order to make it fully compatible with them 2 changes need to happen:
1) There logic that interprets "commands" needs to be abstracted in a way that it works both on actual GraphQL literals and variable values.
2) In case of bulk inserts an additional check needs to be performed to validate that the insert shape is the same across the entire bulk operation.

This is related to #1229

**Comment by 1st1** [comment]
> We use these commands from input objects to compile the query, so these "variables" aren't really query parameters, they may just be the query itself. E.g. OrderUser variable could be passed as {"order": {"name": {"dir": "ASC"}}}

I don't quite like this as it prevents us from pre-compiling GraphQL queries at JS build stage.

I'm also wondering how Prisma / Hasura / Postgraphile / et al handle this? Do they allow this kind of magic?

**Comment by elprans** [comment]
I'm also leaning to restrict variables to data only (i.e no commands).

**Comment by vpetrovykh** [comment]
Restricting variables to data is a lot more vague than it sounds. Are links data? If so, there are two ways a link can be specified during `INSERT` in GraphQL:
1) a new linked object can be specified as purely "data" of the properties of the object
2) an existing object may be linked via a combination of `filter`, `order`, `first`, etc. commands
Considering that required links are not that uncommon, the need to include commands together with "data" is very reasonable.

As far as I can tell using a variable should be indistinguishable from using an equivalent literal in the GraphQL query, so restricting variables is kinda breaking that equivalence. So I would prefer if the restrictions were consistent: i.e. that you could refactor queries. I realize that commands in variables present problems with caching, so for the moment we probably want to disable cache for them. At the end of the day the most likely usage of these fancy variables is the bulk insert scenario, which, I suspect, is not the fasted operation in the first place, so extra translation time is probably not as big of an issue there.

**Comment by 1st1** [comment]
> I realize that commands in variables present problems with caching

Not only that, they are essentially a way to change a query substantially -- add extra filters, nested inserts, etc. I guess for this kind of functionality we'll have forms API some day, but for now we should keep our GraphQL simple.

**Comment by haikyuu** [comment]
+1 for simplicity.

### Links

The first use of Input variables is to insert one data item. If it has links, i would prefer to provide its id than to write a filter function. That's what i do in practice.

Even in EdgeQL, i find it a bit verbose to have to write `owner := SELECT User Filter .name = 'tts' LIMIT 1` rather than `owner := <uuid>'34443'`.

Maybe we should transform the latter to the former at compile time as a convenience.

And if we do that in EdgeQL, mapping it to GraphQL will be straightforward (i suppose).

### Input variables mapping to the query itself

I've never seen a GraphQL API do that before. And it's one of the things that wowed me at first. It's like moving the power of the database to the client!
But in practice, is that really needed?

> We use these commands from input objects to compile the query, so these "variables" aren't really query parameters, they may just be the query itself. E.g. OrderUser variable could be passed as `{"order": {"name": {"dir": "ASC"}}}`. The final query will have an ORDER BY .name ASC clause, but won't have any parameters.

Instead of passing an input variable like `{"order": {"name": {"dir": "ASC"}}}`, i would instead pass `{sortOrder": "ASC"}` and use it as a scalar. It makes the query more readable.

**Comment by tailhook** [comment]
Another example that doesn't work now is:
```
query Name($title: FilterString = {eq: "Hey"}) {
  Text(filter: {title: $title}) {
    id, title
  }
}
```
With quite cryptic error:
```
{
  "errors": [
    {
      "message": "KeyError: 'FilterString'"
    }
  ]
}
```
This is somehow mentioned as "only scalar input variables", just wanted to list it explicitly. Also error reporting on this case have to be improved if we keep variables this way.

**Comment by jnak** [comment]
+1 for this. I'm wondering if anyone has figured some workarounds in the meantime. 

In my case, I'm pretty limited because the input is converted from a scalar variable to an objecttype input by the `graphql-tools/stich` and `graphql-tools/delegate`. I was hoping to use these packages to curate and cleanup the EdgeDB graph so I can neatly abstract away common queries / filtering from my consumers. 

I'm not sure what to do here.... Has anyone looking seriously into making the 2 code changes that @vpetrovykh suggested?
